doc/prebuilt/man/rtcanconfig.1
'\" t
.\"     Title: rtcanconfig
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 2008/04/19
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 2.99.5
.\"  Language: English
.\"
.TH "RTCANCONFIG" "1" "2008/04/19" "Xenomai 2\&.99\&.5" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
rtcanconfig \- Xenomai tool for configuring the CAN controller
.SH "SYNOPSIS"
.sp
\fBrtcanconfig\fR <can\-interface> [options] [up|down|start|stop|sleep]
.SH "DESCRIPTION"
.sp
\fBrtcanconfig\fR is part of Xenomai\&. It is used to configure the CAN controller\&. The system must run a suitable Xenomai enabled kernel with the respective module (CAN)\&.
.SH "OPTIONS"
.sp
\fBrtcanconfig\fR accepts the following options:
.PP
\fB\-v, \-\-verbose\fR
.RS 4
be verbose
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
a usage description
.RE
.PP
\fB\-c, \-\-ctrlmode=CTRLMODE\fR
.RS 4
listenonly, loopback or none
.RE
.PP
\fB\-b, \-\-baudrate=BPS\fR
.RS 4
baudrate in bits/sec
.RE
.PP
\fB\-B, \-\-bittime=BTR0:BTR1\fR
.RS 4
BTR or standard bit\-time
.RE
.sp
\fB\-B, \-\-bittime=BRP:PROP_SEG:PHASE_SEG1:PHASE_SEG2:SJW:SAM\fR
.SH "SEE ALSO"
.sp
\fBrtcanrecv(1)\fR\fB, \fR\fBrtcansend(1)\fR\&.
.SH "AUTHORS"
.sp
\fBrtcanconfig\fR was written by Wolfgang Grandegger and Sebastian Smolorz\&. This man page was written by Roland Stigge\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

utils/can/rtcanconfig.c
/*
 * Program to configuring the CAN controller
 *
 * Copyright (C) 2006 Wolfgang Grandegger <wg@grandegger.com>
 *
 * Copyright (C) 2005, 2006 Sebastian Smolorz
 *                          <Sebastian.Smolorz@stud.uni-hannover.de>
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <errno.h>
#include <getopt.h>
#include <sys/mman.h>

#include <rtdm/can.h>

static void print_usage(char *prg)
{
    fprintf(stderr,
	    "Usage: %s <can-interface> [Options] [up|down|start|stop|sleep]\n"
	    "Options:\n"
	    " -v, --verbose            be verbose\n"
	    " -h, --help               this help\n"
	    " -c, --ctrlmode=CTRLMODE  listenonly, loopback or none\n"
	    " -b, --baudrate=BPS       baudrate in bits/sec\n"
	    " -B, --bittime=BTR0:BTR1  BTR or standard bit-time\n"
	    " -B, --bittime=BRP:PROP_SEG:PHASE_SEG1:PHASE_SEG2:SJW:SAM\n",
	    prg);
}

static can_baudrate_t string_to_baudrate(char *str)
{
    can_baudrate_t baudrate;
    if (sscanf(str, "%i", &baudrate) != 1)
	return -1;
    return baudrate;
}

static int string_to_mode(char *str)
{
    if ( !strcmp(str, "up") || !strcmp(str, "start") )
	return CAN_MODE_START;
    else if ( !strcmp(str, "down") || !strcmp(str, "stop") )
	return CAN_MODE_STOP;
    else if ( !strcmp(str, "sleep") )
	return CAN_MODE_SLEEP;
    return -EINVAL;
}

static int string_to_ctrlmode(char *str)
{
    if ( !strcmp(str, "listenonly") )
	return CAN_CTRLMODE_LISTENONLY;
    else if ( !strcmp(str, "loopback") )
	return CAN_CTRLMODE_LOOPBACK;
    else if ( !strcmp(str, "none") )
	return 0;

    return -1;
}

int main(int argc, char *argv[])
{
    char    ifname[16];
    int     can_fd = -1;
    int     new_baudrate = -1;
    int     new_mode = -1;
    int     new_ctrlmode = 0, set_ctrlmode = 0;
    int     verbose = 0;
    int     bittime_count = 0, bittime_data[6];
    can_baudrate_t *baudrate;
    can_ctrlmode_t *ctrlmode;
    can_mode_t *mode;
    union {
      struct ifreq ifr;
      struct can_bittime bittime;
      can_baudrate_t baudrate;
      can_ctrlmode_t ctrlmode;
      can_mode_t mode;
    } u;
    struct can_bittime *bittime;
    int opt, ret;
    char* ptr;

    struct option long_options[] = {
	{ "help", no_argument, 0, 'h' },
	{ "verbose", no_argument, 0, 'v'},
	{ "baudrate", required_argument, 0, 'b'},
	{ "bittime", required_argument, 0, 'B'},
	{ "ctrlmode", required_argument, 0, 'c'},
	{ 0, 0, 0, 0},
    };

    while ((opt = getopt_long(argc, argv, "hvb:B:c:",
			      long_options, NULL)) != -1) {
	switch (opt) {
	case 'h':
	    print_usage(argv[0]);
	    exit(0);

	case 'v':
	    verbose = 1;
	    break;

	case 'b':
	    new_baudrate = string_to_baudrate(optarg);
	    if (new_baudrate == -1) {
		print_usage(argv[0]);
		exit(0);
	    }
	    break;

	case 'B':
	    ptr = optarg;
	    while (1) {
		bittime_data[bittime_count++] = strtoul(ptr, NULL, 0);
		if (!(ptr = strchr(ptr, ':')))
		    break;
		ptr++;
	    }
	    if (bittime_count != 2 && bittime_count != 6) {
		print_usage(argv[0]);
		exit(0);
	    }
	    break;

	case 'c':
	    ret = string_to_ctrlmode(optarg);
	    if (ret == -1) {
		print_usage(argv[0]);
		exit(0);
	    }
	    new_ctrlmode |= ret;
	    set_ctrlmode = 1;
	    break;

	    break;

	default:
	    fprintf(stderr, "Unknown option %c\n", opt);
	    break;
	}
    }

    /* Get CAN interface name */
    if (optind != argc - 1 && optind != argc - 2) {
	print_usage(argv[0]);
	return 0;
    }

    strncpy(ifname, argv[optind], IFNAMSIZ);
    strncpy(u.ifr.ifr_name, ifname, IFNAMSIZ);

    if (optind == argc - 2) {   /* Get mode setting */
	new_mode = string_to_mode(argv[optind + 1]);
	if (verbose)
	    printf("mode: %s (%#x)\n", argv[optind + 1], new_mode);
	if (new_mode < 0) {
	    print_usage(argv[0]);
	    return 0;
	}
    }

    can_fd = rt_dev_socket(PF_CAN, SOCK_RAW, CAN_RAW);
    if (can_fd < 0) {
	fprintf(stderr, "Cannot open RTDM CAN socket. Maybe driver not loaded? \n");
	return can_fd;
    }

    ret = rt_dev_ioctl(can_fd, SIOCGIFINDEX, &u.ifr);
    if (ret) {
	fprintf(stderr,"Can't get interface index for %s, code = %d\n", ifname, ret);
	return ret;
    }


    if (new_baudrate != -1) {
	if (verbose)
	    printf("baudrate: %d\n", new_baudrate);
	baudrate = &u.baudrate;
	*baudrate = new_baudrate;
	ret = rt_dev_ioctl(can_fd, SIOCSCANBAUDRATE, &u.ifr);
	if (ret) {
	    goto abort;
	}
    }

    if (bittime_count) {
	bittime = &u.bittime;
	if (bittime_count == 2) {
	    bittime->type = CAN_BITTIME_BTR;
	    bittime->btr.btr0 = bittime_data[0];
	    bittime->btr.btr1 = bittime_data[1];
	    if (verbose)
		printf("bit-time: btr0=0x%02x btr1=0x%02x\n",
		       bittime->btr.btr0, bittime->btr.btr1);
	} else {
	    bittime->type = CAN_BITTIME_STD;
	    bittime->std.brp = bittime_data[0];
	    bittime->std.prop_seg = bittime_data[1];
	    bittime->std.phase_seg1 = bittime_data[2];
	    bittime->std.phase_seg2 = bittime_data[3];
	    bittime->std.sjw = bittime_data[4];
	    bittime->std.sam = bittime_data[5];
	    if (verbose)
		printf("bit-time: brp=%d prop_seg=%d phase_seg1=%d "
		       "phase_seg2=%d sjw=%d sam=%d\n",
		       bittime->std.brp,
		       bittime->std.prop_seg,
		       bittime->std.phase_seg1,
		       bittime->std.phase_seg2,
		       bittime->std.sjw,
		       bittime->std.sam);
	}

	ret = rt_dev_ioctl(can_fd, SIOCSCANCUSTOMBITTIME, &u.ifr);
	if (ret) {
	    goto abort;
	}

    }

    if (set_ctrlmode != 0) {
	ctrlmode = &u.ctrlmode;
	*ctrlmode = new_ctrlmode;
	if (verbose)
	    printf("ctrlmode: %#x\n", new_ctrlmode);
	ret = rt_dev_ioctl(can_fd, SIOCSCANCTRLMODE, &u.ifr);
	if (ret) {
	    goto abort;
	}
    }

    if (new_mode != -1) {
	mode = &u.mode;
	*mode = new_mode;
	ret = rt_dev_ioctl(can_fd, SIOCSCANMODE, &u.ifr);
	if (ret) {
	    goto abort;
	}
    }

    rt_dev_close(can_fd);
    return 0;

 abort:
    rt_dev_close(can_fd);
    return ret;
}
